clang/include/clang/CConv/ConstraintsGraph.h

```diff
@@ -12,125 +12,269 @@
 #ifndef _CONSTRAINTSGRAPH_H
 #define _CONSTRAINTSGRAPH_H
 
-#include <boost/graph/graph_traits.hpp>
-#include <boost/graph/adjacency_list.hpp>
 #include "clang/CConv/Constraints.h"
+#include "llvm/ADT/DirectedGraph.h"
+#include "llvm/Support/GraphWriter.h"
+#include "llvm/ADT/GraphTraits.h"
+#include "llvm/ADT/BreadthFirstIterator.h"
 
-using namespace std;
+template<class DataType> struct DataEdge;
 
-template<class G>
-class BaseGraph {
+template<typename DataType, typename EdgeType = DataEdge<DataType>>
```

Comment:
Whoa, I don't understand this syntax. Are you defining/constraining a typename in the middle of your template instantiation? Where can I learn more?

Reply:
C++ lets you have default arguments for template parameters.